### ISSUE TITLE
Prototype configuration-safe outputs for java_library

### ISSUE THREAD

**Issue by gregestren** [bazelbuild/bazel#8339]
This is a sub-issue of #6526 .

The overall project is big, experimental, and touches all kinds of code. This issue represents a first-phase priority of setting up basic infrastructure and getting it to work with an `--experimental` flag on `java_library` compilation. 

The rationale is that there's known value to platform-independent Java outputs and `java_library` doesn't add special complications on top of the core design.

Once we get this working we can expand to other rules.

**Comment by gregestren** [comment]
Expected sequence of initial PRs (these are mostly written and now just being split up into submittable form with appropriate tests):

1. (already submitted): Add --experiment_output_paths=content
1. Provide the ability to declare which artifacts opt into this experiment
1. Add a global map from config-based paths (`bazel-out/x86-fastbuild/...`) to content-based paths (`bazel-out/$(CONTENTDIGEST)/...`), make it available in key places in Bazel
1. Add the ability to content-map paths in action command lines 
1. Add the ability to content-map through one sample executor / object store mode
1. Trigger all of these in java_library as a proof of concept

**Comment by gregestren** [comment]
P1 issue review: still relevant, still very much intend to explore this but I simply am unable to put time into it at the moment. Hoping I can pick this up next quarter.

**Comment by cgruber** [comment]
A year later, any thoughts about this issue, and it's priority?

**Comment by gregestren** [comment]
Yes, with the obvious caveat this issue hasn't been updated recently. Actually I'm surprised - I've pushed through a few PRs recently but I guess I forgot to attach them here:

- https://github.com/bazelbuild/bazel/commit/bb2941bccb0d010223fcbd07139e437b219e837a
- https://github.com/bazelbuild/bazel/commit/526ea392ac50a0f11eb65fd29a6cde5962a08c97
- https://github.com/bazelbuild/bazel/commit/5f9e9f04ecdda8edf779757c7e54aecc29923d91
- https://github.com/bazelbuild/bazel/commit/3ed2c690becb30a08503fb81f52acc4e9831609a

I guess I've been attaching them to https://github.com/bazelbuild/bazel/issues/6526.

Technically this works now on Bazel@head if you add the right flags. I have another pending PR that makes this a bit more straightforward for Java compilation specifically, although still behind an experimental flag. Let me know if you're interested in exploring this further.

After I get that other pending PR in (https://bazel-review.googlesource.com/c/bazel/+/174452) I want to start exploring impact. My tentative experiments on sample projects have shown good results. If you're trying to cache different JDKs (i.e. both Mac and Linux JDKs compiling the same sources), as many people are, there are some extra complications to consider.

**Comment by gregestren** [comment]
FYi I have been making progress on this. PRs have been attached to https://github.com/bazelbuild/bazel/issues/6526 instead of here (I'm not sure that's the right choice, but I guess I've confused myself with the issue overlap). 

As of now Java actions and some Android actions can strip output paths through an experimental flag. That requires cooperation with the executor. I'm doing scale experiments now with Google's executors to get a sense of depot-wide impact.

Note, to be clear, that this speaks to caching within a build or across builds when using homogenous executors. i.e. when every build uses a Linux remote executor or action cache. A related problem is caching actions that could run on both, say, Mac and Linux. The immediate work I'm doing doesn't address that but there is some ongoing discussion about that

**Comment by guw** [comment]
@gregestren Any update on the issue? 

Specifically, we are looking for a Java build solution where a remote cache can be populated by a Linux machine and consumed with cache hits from macOS machines.

**Comment by gregestren** [comment]
@guw  For that purpose I'd look at https://docs.google.com/document/d/1uMPj2s0TlHSIKSngqOkWJoeqOtKzaxQLtBrRfYif3Lo/edit and https://github.com/bazelbuild/bazel/discussions/18378.

**Comment by gregestren** [comment]
New source of truth for this kind of thing is https://github.com/bazelbuild/bazel/discussions/22658. Closing this accordingly.
